File: services/mini-coupang/src/test/java/com/deepblog/minicoupang/domain/product/api/ProductControllerIntegrationTest.java
package com.deepblog.minicoupang.domain.product.api;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.testcontainers.service.connection.ServiceConnection;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.testcontainers.containers.PostgreSQLContainer;
import org.testcontainers.junit.jupiter.Container;
import org.testcontainers.junit.jupiter.Testcontainers;

import java.util.HashMap;
import java.util.Map;

import static org.hamcrest.Matchers.startsWith;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.header;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@SpringBootTest
@AutoConfigureMockMvc
@Testcontainers
class ProductControllerIntegrationTest {

    @Container
    @ServiceConnection
    static PostgreSQLContainer<?> postgres = new PostgreSQLContainer<>("postgres:16");

    @Autowired MockMvc mockMvc;
    @Autowired ObjectMapper objectMapper;

    private String validBody(Map<String, Object> overrides) throws Exception {
        Map<String, Object> base = new HashMap<>();
        base.put("sellerId", 1);
        base.put("sellerName", "Coupang Seller");
        base.put("name", "스마트폰");
        base.put("price", 1_200_000);
        base.put("stock", 10);
        base.putAll(overrides);
        return objectMapper.writeValueAsString(base);
    }

    @Test
    void POST_creates_product_and_returns_201_with_Location() throws Exception {
        mockMvc.perform(post("/api/v1/products")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(validBody(Map.of())))
                .andExpect(status().isCreated())
                .andExpect(header().string("Location", startsWith("/api/v1/products/")))
                .andExpect(jsonPath("$.id").isNumber())
                .andExpect(jsonPath("$.sellerId").value(1))
                .andExpect(jsonPath("$.sellerName").value("Coupang Seller"))
                .andExpect(jsonPath("$.name").value("스마트폰"))
                .andExpect(jsonPath("$.price").value(1_200_000))
                .andExpect(jsonPath("$.stock").value(10));
    }

    @Test
    void GET_existing_returns_200() throws Exception {
        String created = mockMvc.perform(post("/api/v1/products")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(validBody(Map.of("name", "노트북"))))
                .andReturn().getResponse().getContentAsString();
        long id = objectMapper.readTree(created).get("id").asLong();

        mockMvc.perform(get("/api/v1/products/{id}", id))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.name").value("노트북"));
    }

    @Test
    void GET_non_existing_returns_404() throws Exception {
        mockMvc.perform(get("/api/v1/products/999999"))
                .andExpect(status().isNotFound());
    }

    @Test
    void POST_with_blank_name_returns_400() throws Exception {
        mockMvc.perform(post("/api/v1/products")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(validBody(Map.of("name", ""))))
                .andExpect(status().isBadRequest());
    }

    @Test
    void POST_with_negative_price_returns_400() throws Exception {
        mockMvc.perform(post("/api/v1/products")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(validBody(Map.of("price", -100))))
                .andExpect(status().isBadRequest());
    }

    @Test
    void GET_by_sellerId_returns_only_that_seller_products() throws Exception {
        mockMvc.perform(post("/api/v1/products")
                .contentType(MediaType.APPLICATION_JSON)
                .content(validBody(Map.of("sellerId", 77, "name", "TV"))))
                .andExpect(status().isCreated());

        mockMvc.perform(post("/api/v1/products")
                .contentType(MediaType.APPLICATION_JSON)
                .content(validBody(Map.of("sellerId", 77, "name", "Tablet"))))
                .andExpect(status().isCreated());

        mockMvc.perform(post("/api/v1/products")
                .contentType(MediaType.APPLICATION_JSON)
                .content(validBody(Map.of("sellerId", 88, "name", "Other"))))
                .andExpect(status().isCreated());

        mockMvc.perform(get("/api/v1/products").param("sellerId", "77"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.length()").value(2));
    }
}

File: services/product-service/src/test/kotlin/com/deepblog/product/api/ProductControllerIntegrationTest.kt
package com.deepblog.product.api

import com.fasterxml.jackson.databind.ObjectMapper
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.testcontainers.service.connection.ServiceConnection
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.get
import org.springframework.test.web.servlet.post
import org.testcontainers.containers.PostgreSQLContainer
import org.testcontainers.junit.jupiter.Container
import org.testcontainers.junit.jupiter.Testcontainers

@SpringBootTest
@AutoConfigureMockMvc
@Testcontainers
class ProductControllerIntegrationTest {

    companion object {
        @Container
        @ServiceConnection
        @JvmStatic
        val postgres: PostgreSQLContainer<*> = PostgreSQLContainer("postgres:16")
    }

    @Autowired lateinit var mockMvc: MockMvc
    @Autowired lateinit var objectMapper: ObjectMapper

    private fun validBody(overrides: Map<String, Any?> = emptyMap()): String {
        val base = mutableMapOf<String, Any?>(
            "sellerId" to 1,
            "sellerName" to "Coupang Seller",
            "name" to "스마트폰",
            "price" to 1_200_000,
            "stock" to 10,
        )
        base.putAll(overrides)
        return objectMapper.writeValueAsString(base)
    }

    @Test
    fun `POST products creates product and returns 201 with Location`() {
        mockMvc.post("/api/v1/products") {
            contentType = MediaType.APPLICATION_JSON
            content = validBody()
        }.andExpect {
            status { isCreated() }
            header { exists("Location") }
            jsonPath("$.id") { isNumber() }
            jsonPath("$.sellerId") { value(1) }
            jsonPath("$.sellerName") { value("Coupang Seller") }
            jsonPath("$.name") { value("스마트폰") }
            jsonPath("$.price") { value(1_200_000) }
            jsonPath("$.stock") { value(10) }
        }
    }

    @Test
    fun `GET existing product returns 200`() {
        val created = mockMvc.post("/api/v1/products") {
            contentType = MediaType.APPLICATION_JSON
            content = validBody(mapOf("name" to "노트북"))
        }.andReturn().response.contentAsString
        val id = objectMapper.readTree(created).get("id").asLong()

        mockMvc.get("/api/v1/products/$id").andExpect {
            status { isOk() }
            jsonPath("$.name") { value("노트북") }
        }
    }

    @Test
    fun `GET non-existing product returns 404`() {
        mockMvc.get("/api/v1/products/999999").andExpect {
            status { isNotFound() }
        }
    }

    @Test
    fun `POST with blank name returns 400`() {
        mockMvc.post("/api/v1/products") {
            contentType = MediaType.APPLICATION_JSON
            content = validBody(mapOf("name" to ""))
        }.andExpect {
            status { isBadRequest() }
        }
    }

    @Test
    fun `POST with negative price returns 400`() {
        mockMvc.post("/api/v1/products") {
            contentType = MediaType.APPLICATION_JSON
            content = validBody(mapOf("price" to -100))
        }.andExpect {
            status { isBadRequest() }
        }
    }

    @Test
    fun `GET by sellerId returns only that seller's products`() {
        mockMvc.post("/api/v1/products") {
            contentType = MediaType.APPLICATION_JSON
            content = validBody(mapOf("sellerId" to 77, "name" to "TV"))
        }.andExpect { status { isCreated() } }

        mockMvc.post("/api/v1/products") {
            contentType = MediaType.APPLICATION_JSON
            content = validBody(mapOf("sellerId" to 77, "name" to "Tablet"))
        }.andExpect { status { isCreated() } }

        mockMvc.post("/api/v1/products") {
            contentType = MediaType.APPLICATION_JSON
            content = validBody(mapOf("sellerId" to 88, "name" to "Other"))
        }.andExpect { status { isCreated() } }

        mockMvc.get("/api/v1/products?sellerId=77").andExpect {
            status { isOk() }
            jsonPath("$.length()") { value(2) }
        }
    }
}
